Необхідно виконати:

1. Прочитайте csv файл (використовуйте функцію read_csv)
2. Виведіть перші п'ять рядків (використовується функція head)

In [ ]:
import pandas as pd

df = pd.read_csv("bestsellers with categories.csv")
first_5_rows = df.head()
first_5_rows

3. Виведіть розміри датасету (використовуйте атрибут shape)

In [ ]:
df.shape

<div class="alert alert-block alert-info">Про скільки книг зберігає дані датасет?

**Відповідь: 351**</div>

In [ ]:
unique_books = df["Name"].nunique()
print("Про скільки книг зберігає дані датасет? Відповідь:", unique_books)

***
## Первинне дослідження даних
***

In [ ]:
df.columns = ["name", "author", "user_rating", "reviews", "price", "year", "genre"]

Перевірте, чи у всіх рядків вистачає даних: виведіть кількість пропусків (na) у кожному зі стовпців (використовуйте функції isna та sum)

In [ ]:
missing_values = df.isna().sum()
missing_values

<div class="alert alert-block alert-info">Чи є в якихось змінних пропуски? 

**Відповідь: ні**</div>

Перевірте, які є унікальні значення в колонці genre (використовуйте функцію unique)

In [ ]:
unique_genre = df["genre"].unique()
print("Унікальні жанри:", unique_genre)

<div class="alert alert-block alert-info">Які є унікальні жанри? 

**Відповідь: 'Non Fiction', 'Fiction'**</div>

Тепер подивіться на розподіл цін: побудуйте діаграму (використовуйте kind='hist')

In [ ]:
import matplotlib.pyplot as plt

plt.hist(df["price"])
plt.title("The price of the book")
plt.grid(True)
plt.xlabel("price")
plt.ylabel("items")
plt.show

Визначте, яка ціна у нас максимальна, мінімальна, середня, медіанна (використовуйте функції max, min, mean, median)

In [ ]:
max_price = df["price"].max()
min_price = df["price"].min()
av_price = df["price"].mean()
median_price = df["price"].median()

max_price, min_price, av_price, median_price


<div class="alert alert-block alert-info"> Максимальна ціна?    

 **Відповідь: 105**  
 
Мінімальна ціна?    
 **Відповідь: 0** 

Середня ціна?   
 **Відповідь: 13.1** 

Медіанна ціна?  
 **Відповідь: 11.0**
</div>


***
## Пошук та сортування даних
***

In [ ]:
max_rating = df["user_rating"].max()
max_rating

<div class="alert alert-block alert-info"> Відповідь: Який рейтинг у датасеті найвищий?   

 **Відповідь: 4.9**  
</div>

In [ ]:
max_rating_books = df[df["user_rating"]==max_rating]["name"].nunique()
number_of_max_rating_books = max_rating_books
number_of_max_rating_books

<div class="alert alert-block alert-info"> Скільки книг мають такий рейтинг?   

 **Відповідь: 28**  
</div>

In [ ]:
book_with_max_reviews = df[df["reviews"]==df["reviews"].max()]["name"]
book_with_max_reviews

<div class="alert alert-block alert-info"> Яка книга має найбільше відгуків?   

 **Відповідь: Where the Crawdads Sing**  
</div>

In [82]:
top_50_2015 = df[df["year"] == 2015]
most_expensive_book_2015 = top_50_2015[
    top_50_2015["price"] == top_50_2015["price"].max()
]["name"]
most_expensive_book_2015

277    Publication Manual of the American Psychologic...
Name: name, dtype: object

<div class="alert alert-block alert-info"> З тих книг, що потрапили до Топ-50 у 2015 році, яка книга найдорожча (можна використати проміжний датафрейм)?

 **Відповідь:  Publication Manual of the American Psychologic...**  
</div>

In [86]:
top_2010_fiction = df[(df["year"]==2010) & (df["genre"] == "Fiction")]
len(top_2010_fiction)

20

<div class="alert alert-block alert-info"> Скільки книг жанру Fiction потрапили до Топ-50 у 2010 році (використовуйте &)?

 **Відповідь:  20**  
</div>

In [87]:
books_with_high_rating_isin = df[
    (df["user_rating"] == 4.9) & (df["year"].isin([2010, 2011]))
]

books_with_high_rating = df[
    (df["user_rating"] == 4.9) & ((df["year"] == 2010) | (df["year"] == 2011))
]

print(len(books_with_high_rating_isin), len(books_with_high_rating))

1 1


<div class="alert alert-block alert-info"> Скільки книг з рейтингом 4.9 потрапило до рейтингу у 2010 та 2011 роках (використовуйте | або функцію isin)?

 **Відповідь:  1**  
</div>

І насамкінець, давайте відсортуємо за зростанням ціни всі книги, які потрапили до рейтингу в 2015 році і коштують дешевше за 8 доларів (використовуйте функцію sort_values).

In [92]:
books_2015_under_8 = df[(df["year"] == 2015) & (df["price"] < 8)].sort_values(
    by="price", ascending=True
)
last_book_name = books_2015_under_8.iloc[-1]["name"]
last_book_name

'Old School (Diary of a Wimpy Kid #10)'

<div class="alert alert-block alert-info"> Яка книга остання у відсортованому списку?

 **Відповідь:  'Old School (Diary of a Wimpy Kid #10)'**  
</div>

***
## Агрегування даних та з'єднання таблиць
***

In [94]:
price_stats_by_genre = df.groupby("genre")["price"].agg(
    min_price="min", max_price="max"
)
price_stats_by_genre

min_price  max_price
genre                            
Fiction              0         82
Non Fiction          0        105

<div class="alert alert-block alert-info"> Максимальна ціна для жанру Fiction:    

 **Відповідь: 82**  
 
Мінімальна ціна для жанру Fiction?    
 **Відповідь: 0** 

Максимальна ціна для жанру Non Fiction?   
 **Відповідь: 105** 

Мінімальна ціна для жанру Non Fiction?  
 **Відповідь: 0**
</div>


Тепер створіть новий датафрейм, який вміщатиме кількість книг для кожного з авторів (використовуйте функції groupby та agg, для підрахунку кількості використовуйте count). Не беріть усі стовпці, виберете тільки потрібні

In [99]:
books_by_author = (
    df.groupby("author")["name"].agg(book_count="count").sort_values(by="book_count")
)
print(books_by_author)
books_by_author.shape

                                    book_count
author                                        
Jill Twiss                                   1
Patrick Thorpe                               1
Ian K. Smith M.D.                            1
Howard Stern                                 1
Hopscotch Girls                              1
...                                        ...
American Psychological Association          10
Suzanne Collins                             11
Gary Chapman                                11
Rick Riordan                                11
Jeff Kinney                                 12

[248 rows x 1 columns]


(248, 1)

<div class="alert alert-block alert-info"> Якої розмірності вийшла таблиця?

 **Відповідь:  (248, 1)**  

 Який автор має найбільше книг?     
 **Відповідь:  Jeff Kinney** 

Скільки книг цього автора?      
 **Відповідь:  12**

</div>

Тепер створіть другий датафрейм, який буде вміщати середній рейтинг для кожного автора (використовуйте функції groupby та agg, для підрахунку середнього значення використовуйте mean). Не беріть усі стовпці, виберете тільки потрібні

In [103]:
av_rating_by_author = (
    df.groupby("author")["user_rating"]
    .agg(avg_rating="mean")
    .sort_values(by="avg_rating")
)

av_rating_by_author

avg_rating
author                        
Donna Tartt                3.9
Gillian Flynn              4.0
Muriel Barbery             4.0
Gallup                     4.0
Ian K. Smith M.D.          4.1
...                        ...
Lin-Manuel Miranda         4.9
Patrick Thorpe             4.9
Nathan W. Pyle             4.9
Bill Martin Jr.            4.9
Jill Twiss                 4.9

[248 rows x 1 columns]

<div class="alert alert-block alert-info"> У якого автора середній рейтинг мінімальний?

 **Відповідь:  Donna Tartt**  

 Який у цього автора середній рейтинг?     
 **Відповідь:  3.9** 
</div>

З'єднайте останні два датафрейми так, щоб для кожного автора було видно кількість книг та середній рейтинг (Використовуйте функцію concat з параметром axis=1). Збережіть результат у змінну

In [105]:
author_stats = pd.concat([books_by_author, av_rating_by_author], axis=1)
author_stats

book_count  avg_rating
author                                                    
Jill Twiss                                   1    4.900000
Patrick Thorpe                               1    4.900000
Ian K. Smith M.D.                            1    4.100000
Howard Stern                                 1    4.300000
Hopscotch Girls                              1    4.800000
...                                        ...         ...
American Psychological Association          10    4.500000
Suzanne Collins                             11    4.663636
Gary Chapman                                11    4.736364
Rick Riordan                                11    4.772727
Jeff Kinney                                 12    4.800000

[248 rows x 2 columns]

Відсортуйте датафрейм за зростаючою кількістю книг та зростаючим рейтингом (використовуйте функцію sort_values)

In [109]:
author_stats.sort_values(by=["book_count", "avg_rating"], ascending= [True, True])

book_count  avg_rating
author                                                    
Muriel Barbery                               1    4.000000
Ian K. Smith M.D.                            1    4.100000
Pierre Dukan                                 1    4.100000
Chris Cleave                                 1    4.100000
Geneen Roth                                  1    4.200000
...                                        ...         ...
American Psychological Association          10    4.500000
Suzanne Collins                             11    4.663636
Gary Chapman                                11    4.736364
Rick Riordan                                11    4.772727
Jeff Kinney                                 12    4.800000

[248 rows x 2 columns]

<div class="alert alert-block alert-info"> Який автор перший у списку?

 **Відповідь:  Muriel Barbery**  
</div>